sympy/polys/polytools.py

```diff
@@ -3773,6 +3775,174 @@ def nth_power_roots_poly(f, n):
+
+        Returns
+        =======
+
+        :py:class:`~.Expr`
+            A non-negative lower bound which will be positive if all the roots
+            are distinct.
+
+        Raises
+        ======
+
+        DomainError
+            If the domain of the polynomial is not :ref:`ZZ` or :ref:`QQ`.
+        MultivariatePolynomialError
+            If the polynomial is not univariate.
+
+        """
+        dom = f.get_domain()
+        if not (dom.is_IntegerRing or dom.is_RationalField):
+            raise DomainError("Only polys over QQ and ZZ are supported.")
+        if f.is_multivariate:
+            raise MultivariatePolynomialError(
+                "Must be a univariate polynomial")
+        if f.is_monomial:
+            return Integer(0)
+
+        x = f.gen
+        y = Dummy('y')
+        g = f.subs(x, x + y)
+        h = Poly(resultant(f, g, gens=[x, y]), domain=f.domain)
+        # If f has roots ri, the roots of h are all the ri - rj.
+
+        if h.is_monomial:
+            # In this case all roots of f are the same.
+            return Integer(0)
+
+        lb, _ = h.root_bounds()
+
+        # Since h is an even function, the square root of lb is actually a
+        # lower bound on the separation of roots. So lb is a lower bound on
+        # the squared separation.
+        return lb
+
+    def root_equality_test(f):
+        """
+        Returns a binary function that efficiently decides whether two roots of
+        this polynomial are equal.
+
+        Examples
+        ========
+
+        >>> from sympy import Poly, cyclotomic_poly, exp, I, pi
+        >>> f = Poly(cyclotomic_poly(5))
+        >>> eq = f.root_equality_test()
+        >>> r0 = exp(2*I*pi/5)
+        >>> indices = [i for i, r in enumerate(f.all_roots()) if eq(r, r0)]
+        >>> print(indices)
+        [3]
+
+        Raises
+        ======
+
+        DomainError
+            If the domain of the polynomial is not :ref:`ZZ` or :ref:`QQ`.
+        MultivariatePolynomialError
+            If the polynomial is not univariate.
+
+        """
+        delta_sq = f.root_separation_lower_bound_squared()
+        # We have delta_sq = delta**2, where delta is a lower bound on the
+        # minimum separation between any two roots of this polynomial.
+        # Let eps = delta/3, and define eps_sq = eps**2 = delta**2/9.
+        eps_sq = delta_sq / 9
+
+        r, _, _, _ = evalf(1/eps_sq, 1, {})
+        n = fastlog(r)
+        # Then 2^n > 1/eps**2.
+        m = math.ceil(n/2)
+        # Then 2^(-m) < eps.
+        ev = lambda x: quad_to_mpmath(evalf_with_bounded_error(x, m=m))
+
+        # Then for any complex numbers a, b we will have
+        # |a - ev(a)| < eps and |b - ev(b)| < eps.
+        # So if |ev(a) - ev(b)|**2 < eps**2, then
+        # |ev(a) - ev(b)| < eps, hence |a - b| < 3*eps = delta.
+        def eq_test(a, b):
+            A, B = ev(a), ev(b)
+            return (A.real - B.real)**2 + (A.imag - B.imag)**2 < eps_sq
+
+        return eq_test
```

Comment:
Returning the function like this seems like an odd interface

Reply:
I guess since the `Poly` is considered immutable it could just store the test function as an attribute, and return only the True/False answer to the user.

Reply:
But now that I take a closer look, I see that the `Poly` class is actually really minimal in its instance attributes. It only has `rep` and `gens`, and it declares these as its slots `__slots__ = ('rep', 'gens')`. I'm not sure I want to disturb this....

Reply:
Is there much value in storing anything?

Reply:
If there is then it could be possible to have a cached function that returns the `eq_test` but still expose an interface to the user that looks more natural like `p.same_roots(r1, r2)`.

Reply:
> Is there much value in storing anything?

I was running some timing tests in order to try to answer this, when I realized it's not scaling as well as I had hoped. I should have done these scaling tests much earlier!

I want to try `dup_zz_mignotte_bound` and see if it will do better. ([Mignotte bound](https://en.wikipedia.org/wiki/Geometrical_properties_of_polynomial_roots#Root_separation))

Reply:
Couple of findings: (a) `dup_zz_mignotte_bound` is actually a different Mignotte bound, but anyway (b) the Mignotte bound I meant isn't doing much better than the bound we already have.

Might have to think some more...

Reply:
Mignotte bound is currently misnamed.

Reply:
What exactly is slow?

Reply:
> What exactly is slow?

I'm looking at cyclotomic polynomials `phi_n`. I didn't anticipate how fast the coefficients of `Res(phi_n(x), phi_n(x + y))` would grow with `n`.

Of course for cyclotomic polys the geometry is easy, so better separation bounds could be used.

Anyway, the comparison with Mignotte is interesting. Our bound starts out several orders of magnitude better, but then gets worse. As for the run time, we're slowing down dramatically but Mignotte barely changes at all.


```python
f = lambda n: Poly(cyclotomic_poly(prime(n)))

def mignotte(f):
    n = f.degree()
    D = f.discriminant()
    eucl_norm = sqrt( sum( [cf**2 for cf in f.all_coeffs()] ) )
    denom = sqrt(n)**(n+1) * eucl_norm**(n-1)
    num = sqrt(3*abs(D))
    return num/denom

for n in range(2, 11):
    t = time.time()
    b1 = sqrt(f(n).root_separation_lower_bound_squared()).evalf()
    t1 = time.time() - t

    t = time.time()
    b2 = mignotte(f(n)).evalf()
    t2 = time.time() - t
    print(f'{prime(n)}:\n{b1}, {b2}')
    print(f'{t1:.5f}, {t2:.5f}\n')

3:
0.866025403784439, 0.612372435695794
0.00404, 0.00108

5:
0.845154254728517, 0.0541265877365274
0.00793, 0.00131

7:
0.509175077217316, 0.00327364250549328
0.01933, 0.00160

11:
0.00671827668354308, 0.00000547722557505166
0.11216, 0.00210

13:
0.000155324265551629, 1.67448988340192E-7
0.28852, 0.00230

17:
2.09849711164693E-9, 1.00818626091867E-10
1.15808, 0.00292

19:
1.82097216957742E-12, 2.05812538301822E-12
2.17408, 0.00345

23:
2.41307054558417E-20, 6.31974853510620E-16
6.41855, 0.00444

29:
4.75406277160334E-36, 1.79791507099308E-21
24.15388, 0.00471

```